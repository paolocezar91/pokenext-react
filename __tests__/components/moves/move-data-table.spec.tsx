import MoveDataTable from "@/components/moves/move-data-table";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { Move } from "pokeapi-typescript";

// Mock dependencies
jest.mock("next/image", () => ({
  __esModule: true,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  default: (props: any) => <img {...props} />, // Simplified Image mock
}));

jest.mock("@/context/user-context", () => ({
  useUser: () => ({
    settings: { typeArtworkUrl: "sword-shield" },
  }),
}));

jest.mock("next-intl", () => ({
  useTranslations: () => (key: string) => key,
  useLocale: () => "en",
}));

describe("MoveDataTable Component", () => {
  const mockMoveData: Move = {
    name: "thunderbolt",
    type: {
      name: "electric",
      url: "https://pokeapi.co/api/v2/type/13/",
    },
    damage_class: {
      name: "special",
      url: "",
    },
    power: 90,
    accuracy: 100,
    pp: 15,
    // Include other required Move properties with mock values
    id: 1,
    effect_chance: 0,
    effect_entries: [],
    flavor_text_entries: [],
    generation: { name: "", url: "" },
    names: [],
    past_values: [],
    stat_changes: [],
    target: { name: "", url: "" },
  } as unknown as Move;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render the table with all move data", () => {
    render(<MoveDataTable moveData={mockMoveData} />);

    // Verify table title
    expect(
      screen.getByRole("heading", { name: "moves.moveData.title" })
    ).toBeInTheDocument();

    // Verify all headers are rendered
    expect(screen.getByText("pokedex.details.moves.type")).toBeInTheDocument();
    expect(screen.getByText("pokedex.details.moves.class")).toBeInTheDocument();
    expect(screen.getByText("pokedex.details.moves.power")).toBeInTheDocument();
    expect(
      screen.getByText("pokedex.details.moves.accuracy")
    ).toBeInTheDocument();
    expect(screen.getByText("pokedex.details.moves.pp")).toBeInTheDocument();

    // Verify type image
    const typeImage = screen.getByAltText("electric");
    expect(typeImage).toHaveAttribute(
      "src",
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/types/generation-viii/sword-shield/13.png"
    );

    // Verify damage class tooltip and image
    const classImage = screen.getByAltText("special");
    expect(classImage).toHaveAttribute("src", "/move-special.png");

    // Verify numeric values
    expect(screen.getByText("90")).toBeInTheDocument();
    expect(screen.getByText("100")).toBeInTheDocument();
    expect(screen.getByText("15")).toBeInTheDocument();
  });

  it("should handle null power value", () => {
    const moveDataWithoutPower = {
      ...mockMoveData,
      power: null,
    } as unknown as Move;
    render(<MoveDataTable moveData={moveDataWithoutPower} />);
    expect(screen.getByText("-")).toBeInTheDocument();
  });

  it("should handle null accuracy value", () => {
    const moveDataWithoutAccuracy = {
      ...mockMoveData,
      accuracy: null,
    } as unknown as Move;
    screen.debug();
    render(<MoveDataTable moveData={moveDataWithoutAccuracy} />);
    expect(screen.getByText("-")).toBeInTheDocument();
  });

  it("should handle null pp value", () => {
    const moveDataWithoutPP = {
      ...mockMoveData,
      pp: null,
    } as unknown as Move;
    render(<MoveDataTable moveData={moveDataWithoutPP} />);
    expect(screen.getByText("-")).toBeInTheDocument();
  });

  it("should match snapshot with complete data", () => {
    const { asFragment } = render(<MoveDataTable moveData={mockMoveData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot with null values", () => {
    const moveDataWithNulls = {
      ...mockMoveData,
      power: null,
      accuracy: null,
      pp: null,
    } as unknown as Move;
    const { asFragment } = render(
      <MoveDataTable moveData={moveDataWithNulls} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
