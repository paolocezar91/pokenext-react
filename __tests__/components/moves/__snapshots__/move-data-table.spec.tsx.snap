// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveDataTable Component should match snapshot with complete data 1`] = `
<DocumentFragment>
  <div
    class="move-data mt-2 w-full"
  >
    <h3
      class="w-fit text-lg mb-4"
    >
      moves.moveData.title
    </h3>
    <table
      class="w-full"
    >
      <tbody>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.type
          </th>
          <td
            class="py-1"
          >
            <a
              href="/type/electric"
            >
              <img
                alt="electric"
                height="20"
                loading="lazy"
                src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/types/generation-viii/sword-shield/13.png"
                width="100"
              />
            </a>
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.class
          </th>
          <td
            class="py-1"
          >
            <div
              class="tooltip-container"
            >
              <div
                class="tooltip-trigger"
              >
                <span
                  class="flex items-center"
                >
                  <img
                    alt="special"
                    class="mr-2"
                    height="35"
                    loading="lazy"
                    src="/move-special.png"
                    width="35"
                  />
                  Special
                </span>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.power
          </th>
          <td
            class="py-1 w-50"
          >
            90
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.accuracy
          </th>
          <td
            class="py-1 w-50"
          >
            100
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.pp
          </th>
          <td
            class="py-1 w-50"
          >
            15
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`MoveDataTable Component should match snapshot with null values 1`] = `
<DocumentFragment>
  <div
    class="move-data mt-2 w-full"
  >
    <h3
      class="w-fit text-lg mb-4"
    >
      moves.moveData.title
    </h3>
    <table
      class="w-full"
    >
      <tbody>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.type
          </th>
          <td
            class="py-1"
          >
            <a
              href="/type/electric"
            >
              <img
                alt="electric"
                height="20"
                loading="lazy"
                src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/types/generation-viii/sword-shield/13.png"
                width="100"
              />
            </a>
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.class
          </th>
          <td
            class="py-1"
          >
            <div
              class="tooltip-container"
            >
              <div
                class="tooltip-trigger"
              >
                <span
                  class="flex items-center"
                >
                  <img
                    alt="special"
                    class="mr-2"
                    height="35"
                    loading="lazy"
                    src="/move-special.png"
                    width="35"
                  />
                  Special
                </span>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.power
          </th>
          <td
            class="py-1 w-50"
          >
            -
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.accuracy
          </th>
          <td
            class="py-1 w-50"
          >
            -
          </td>
        </tr>
        <tr>
          <th
            class="text-left"
          >
            pokedex.details.moves.pp
          </th>
          <td
            class="py-1 w-50"
          >
            -
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
